### PR TITLE
JENKINS-65033 Run / Project summary - tables to divs

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-65033](https://issues.jenkins.io/browse/JENKINS-65033).

This required some minor adaption in a few plugins that had a block layout (should have been inline) next to the icon, see https://github.com/jenkinsci/warnings-ng-plugin/pull/841, https://github.com/jenkinsci/code-coverage-api-plugin/pull/185

This is a lot safer than forms as it won't break anything, possibly minor changes in some plugins, I've tested all the ones we use on ci.jenkins.io and https://plugins.jenkins.io/sauce-ondemand which was broken without this change due to tables nesting

### Proposed changelog entries

* Entry 1: JENKINS-65033, Run summary page now uses div for layout
* Entry 2: Developer: `divBasedRunLayout` jelly variable available for detecting if Jenkins uses divs for run summary page
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:307]: used in https://github.com/jenkinsci/workflow-job-plugin/pull/180 to adapt to this change without requiring a core bump

**Review by uhafner — APPROVED** [review]
Since most of those forms (`summary.jelly`) start with a copy and paste operation I assume that there are a lot of additional plugins that need to be changed.

**Comment by daniel-beck** [comment]
Could we hold this a bit until we've closed out the previous t2d migration? It's barely in LTS, so I expect further regression reports etc.

**Comment by res0nance** [comment]
> Could we hold this a bit until we've closed out the previous t2d migration? It's barely in LTS, so I expect further regression reports etc.

I agree with Daniel, there is likely to be more regressions.

Also there seems to be a failing test.

**Comment by timja** [comment]
> > Could we hold this a bit until we've closed out the previous t2d migration? It's barely in LTS, so I expect further regression reports etc.

> I agree with Daniel, there is likely to be more regressions.

the scale for this is far smaller, we can wait a bit but doesn't need a year like the other one

> Also there seems to be a failing test.

Yes aware, my IDE doesn't want to run the test without throwing nullpointers in unrelated places -.-

File: core/src/main/resources/hudson/model/CauseAction/summary.jelly
Comment (timja) [inline core/src/main/resources/hudson/model/CauseAction/summary.jelly:28]: needed to add something to go next to the icon to make it looks okay,

any better wording or design thoughts?

image

image

Comment (res0nance) [reply core/src/main/resources/hudson/model/CauseAction/summary.jelly:28]: Maybe a more explicit `Build started by` or `Build caused by`? `Caused by` seems fine as well

Not sure whats a good wording for it is.

Personally I'm not a fan of the icon but that's outside the scope of this PR and with the new icons coming in this might be updated at some point.

**Review by fqueiruga — APPROVED** [review]
Code looks good. Same as above, I'd still suggest having a LTS line stable before going ahead with this.

File: core/src/main/resources/hudson/model/AbstractBuild/index.jelly
Comment (rsandell) [inline core/src/main/resources/hudson/model/AbstractBuild/index.jelly:110]: I'm guessing quite a lot more of breaking in plugins.
Ex: https://github.com/jenkinsci/build-failure-analyzer-plugin/blob/master/src/main/resources/com/sonyericsson/jenkins/plugins/bfa/model/FailureCauseBuildAction/summary.groovy

Comment (timja) [reply core/src/main/resources/hudson/model/AbstractBuild/index.jelly:110]: Pretty sure I tested BFA and it rendered fine, can double check later

File: core/src/main/resources/hudson/model/CauseAction/summary.jelly
Comment (jglick) [reply core/src/main/resources/hudson/model/CauseAction/summary.jelly:28]: Looks fine to me.

**Review by jglick — APPROVED** [review]

File: core/src/main/resources/hudson/model/AbstractBuild/index.jelly
Comment (timja) [reply core/src/main/resources/hudson/model/AbstractBuild/index.jelly:110]: Job page, browser allows the table and it looks the same as before:

image

Run page, browser throws out the table tags, it looks fine except it's using a heading tag which is defined as a block element

image

We could maybe set an adjacent sibling selector to make headings display inline if after an `img` tag or svg, any thoughts @fqueiruga ?

Doesn't look like a blocker anyway to me, thoughts?

**Review by oleg-nenashev — APPROVED** [review]
It will likely lead to some UI regressions, and it is highly likely we will need upgrade guidelines for that. Besides that, I think this is ready to go.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

File: core/src/main/resources/hudson/model/AbstractBuild/index.jelly
Comment (rsandell) [reply core/src/main/resources/hudson/model/AbstractBuild/index.jelly:110]: Nah, that looks OK. Hopefully it won't break things rendered after it, though I don't think it will since it closes it's table tags.

File: core/src/main/resources/hudson/model/AbstractBuild/index.jelly
Comment (rsandell) [inline core/src/main/resources/hudson/model/AbstractBuild/index.jelly:59]: Not sure how things are done with styling etc now in the modern times. But I think it would be nice if this div had an `id` for page decorators et.al. if they need to manipulate it somehow. Themes would also have it easier to style it I guess?

File: core/src/main/resources/hudson/model/AbstractProject/main.jelly
Comment (rsandell) [inline core/src/main/resources/hudson/model/AbstractProject/main.jelly:33]: maybe this one too?

**Review by rsandell — APPROVED** [review]

File: core/src/main/resources/hudson/model/AbstractBuild/index.jelly
Comment (timja) [reply core/src/main/resources/hudson/model/AbstractBuild/index.jelly:59]: not really sure why you would need them.

themes should be able to use selectors and if they are finding a need for it they can add them at the time.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
A justification for a potential breaking change should be provided. None is given in this PR, or at least it's unclear: The description indicates that this fixes a plugin's UI. Was it always broken and this just makes it work (at what cost?)? Did we cause a regression (where?) and this is a regression fix?

Needs corresponding developer documentation.

I would also like to see evidence of a more exhaustive ecosystem search for incompatible plugins before this is merged, or an explanation why this would be an excessively rare occurrence.

Maybe I'm overly cautious here but we've had many regressions from config form t2d, some right here in core and in extremely popular plugins that were only discovered after merging, and I would not want to see this again.

**Comment by daniel-beck** [comment]
I found a few plugins that show a `<t:pane>` ( :trollface: ), which is a `<div>`, inside a `<t:summary>`. Is that a problem? I can't tell from the description of this PR and the examples provided.

**Comment by timja** [comment]
> A justification for a potential breaking change should be provided. 

1. Uses proper layout techniques, so browsers don't try fix markup if anyone does anything it doesn't like, table has very specific elements that are allowed as children and browsers will 'correct' your mistake often changing the pages in ways you don't want.
2. Fixed the sauce-ondemand plugin action which looks ok if it's the only action which is what the authors will have tested, but as you soon as you attach another action, say the test results from junit, and the page is wildly broken. Likely always broken or changed in browsers ages ago.

> I found a few plugins that show a `<t:pane>`, which is a `<div>`, inside a `<t:summary>`. Is that a problem? I can't tell from the description of this PR and the examples provided.

I would expect those plugins to be troublesome already, they could already break the layout as that sounds like the scenario this PR is trying to fix.

What plugins?

**Comment by daniel-beck** [comment]
> I would expect those plugins to be troublesome already, they could already break the layout as that sounds like the scenario this PR is trying to fix. … What plugins?

https://plugins.jenkins.io/groovy-label-assignment/ and https://plugins.jenkins.io/show-build-parameters/ have `:pane`s.

https://plugins.jenkins.io/violations/ has both a `<table>` and a `<div>` inside a `t:summary`. https://plugins.jenkins.io/allure-jenkins-plugin/ has a `<table>`. https://plugins.jenkins.io/appetize/ has a `<div>`.

If `<ul>` is also a problem, then there's https://plugins.jenkins.io/cccc/ (see screenshot provided) and https://plugins.jenkins.io/buildcontext-capture/

**Comment by timja** [comment]
`ul` is fine, (it's used in this PR)

Markup here looks really weird, the same outcome would've been done with no table or anything =/, unless i'm missing something: https://github.com/jenkinsci/allure-plugin/blob/master/src/main/resources/ru/yandex/qatools/allure/jenkins/AllureBuildAction/summary.jelly#L3-L7, it would need checking, I can't tell from looking at it if it would render ok.

the pane ones are worth checking, the title will likely be moved onto a new line rather than adjacent to the image

Other than misunderstanding of html layout this will work: https://plugins.jenkins.io/appetize/, instead of `br`, `div` or `p` should be used to move it on the next line.

**Comment by timja** [comment]
> @timja do you wish and have time to revive this PR as you said on #6307? At this point, the PR seems stalled or could be close. Sorry if I'm wrong.

it will get done at some point, please leave it alone for now

**Comment by timja** [comment]
No plans to work on this atm

**Pull request closed without merging**
